stylelint-polaris/plugins/coverage/index.js

```diff
@@ -2,79 +2,114 @@ const stylelint = require('stylelint');
+ */
+
+/**
+ * @typedef {object} CategorySettings
+ * @property {import('stylelint').RuleMessage} [message] - Message appended to the warning if no custom message is set on a rule's secondary options
+ * @property {import('stylelint').RuleMeta} [meta] - Category documentation URL hyperlinked to the reported rule in the VS Code diagnostic
+ */
 
 /**
  * @typedef {{
- *   [category: string]: import('stylelint').ConfigRules
- * }} PrimaryOptions
+ *   [category: string]: StylelintRuleConfig | [
+ *     StylelintRuleConfig, CategorySettings
+ *   ]
+ * }} CategorizedRules
  */
 
 // Setting `line` to an invalid line number forces the warning to be reported
 // and the `report({node})` option is used to display the location information:
 // https://github.com/stylelint/stylelint/blob/57cbcd4eb0ee809006a1e3d2ccfe73af48744ad5/lib/utils/report.js#L49-L52
 const forceReport = {line: -1};
+const defaultMeta = {
+  url: 'https://github.com/Shopify/polaris/tree/main/stylelint-polaris/plugins/coverage/README.md#rules',
+};
 
 module.exports = stylelint.createPlugin(
-  ruleName,
-  /** @param {PrimaryOptions} primaryOptions */
-  (primaryOptions, secondaryOptions, context) => {
-    const isPrimaryOptionsValid = validatePrimaryOptions(primaryOptions);
-
-    const rules = [];
-
-    for (const [categoryName, categoryConfigRules] of Object.entries(
-      primaryOptions,
-    )) {
-      for (const [categoryRuleName, categoryRuleSettings] of Object.entries(
-        categoryConfigRules,
+  coverageRuleName,
+  /**
+   * @param {CategorizedRules} categorizedRules - Configured Stylelint rules grouped by Polaris coverage category
+   * @param {import('stylelint').RuleContext} context
+   */
+  (categorizedRules, _, context) => {
+    const isPrimaryOptionsValid = validatePrimaryOptions(categorizedRules);
+    const coverageRules = [];
+
+    for (const [category, categoryConfig] of Object.entries(categorizedRules)) {
+      const [stylelintRules, categorySettings] =
+        normalizeConfig(categoryConfig);
+
+      for (const [stylelintRuleName, stylelintRuleConfig] of Object.entries(
+        stylelintRules,
       )) {
-        rules.push({
-          coverageRuleName: `polaris/${categoryName}/${categoryRuleName}`,
-          categoryRuleName,
-          categoryRuleSettings,
-          categoryRuleSeverity: categoryRuleSettings?.[1]?.severity,
-          categoryRuleFix:
-            context.fix && !categoryRuleSettings?.[1]?.disableFix,
+        coverageRules.push({
+          ruleName: `polaris/${category}/${stylelintRuleName}`,
+          stylelintRuleName,
+          ruleSettings: stylelintRuleConfig,
+          severity: stylelintRuleConfig?.[1]?.severity,
+          fix: context.fix && !stylelintRuleConfig?.[1]?.disableFix,
+          appendedMessage:
+            stylelintRuleConfig?.[1]?.message || categorySettings?.message,
+          meta: categorySettings?.meta || defaultMeta,
```

Comment:
**question:** Is it possible to allow individual rules to define their own `meta`? Might be neat in some cases to provide links directly to the `README.md#<category>/<rule-name>`